### PR TITLE
Feature/school 485/add teacher

### PR DESCRIPTION
**Pull request by Feazom**

### CODE REVIEW

File: open_schools_platform/organization_management/teachers/services.py
Comment (Ivan228777) [inline open_schools_platform/organization_management/teachers/services.py:16]: Thing about creating teacher profile, do we need to create it automatically, when we are creating user (as other profiles)? If we do need that, you are going to need to add create_teacher_profile to create_user service.

File: open_schools_platform/organization_management/teachers/services.py
Comment (Ivan228777) [inline open_schools_platform/organization_management/teachers/services.py:58]: I have a question, am I right, that this query handler is for situations, where TeacherProfile sends query to circle, so circle can hire him as a teacher? Did product approve this scenario? If he did, then you can just resolve this conversation, we are going to need a second query handler for situations, where circle actually sends query to TeacherProfile but we will create it later.

**Review by Ivan228777 — CHANGES_REQUESTED** [review]

File: open_schools_platform/organization_management/teachers/services.py
Comment (Feazom) [reply open_schools_platform/organization_management/teachers/services.py:58]: No it should be Cirlce send query to TeacherProfile. Product approve that/

Comment (Ivan228777) [reply open_schools_platform/organization_management/teachers/services.py:58]: So product didn't approved that TeacherProfile can send query to Circle?
Because this query_handler can be used only when TeacherProfile sends query to Circle because it gets circle_access from query.recipient. If you wanted to make this query_handler for cases, where Circle sends query to TeacherProfile, then you need to get circle_access from query.sender and teacher_profile_access from query.recipient.

Correct me if I didn't get your idea.

**Review by Ivan228777 — CHANGES_REQUESTED** [review]
Nice, you can fix query_handler and I will approve.

File: open_schools_platform/organization_management/circles/services.py
Comment (Ivan228777) [inline open_schools_platform/organization_management/circles/services.py:97]: You need to refactor these lines, because they doesn't work if TeacherProfile is recipient and Circle is sender.

For example, here TeacherProfile can only set canceled status, but this is wrong, because he is the one who received this query. So he need to have ability to accept this query or decline it. 

You can see the above query_handler (for Family), so you can take logic from there as a reference.

**Review by Ivan228777 — CHANGES_REQUESTED** [review]

**Review by zavarin-michael — CHANGES_REQUESTED** [review]

File: open_schools_platform/organization_management/circles/tests/views/test_circle_invite_teacher.py
Comment (Ivan228777) [inline open_schools_platform/organization_management/circles/tests/views/test_circle_invite_teacher.py:70]: ^^

Comment (Ivan228777) [reply open_schools_platform/organization_management/circles/tests/views/test_circle_invite_teacher.py:70]: You are creating such teacher in every test, so I won't point it out down below, but, please, check them as well.

File: open_schools_platform/organization_management/circles/tests/views/test_circle_invite_teacher.py
Comment (Ivan228777) [inline open_schools_platform/organization_management/circles/tests/views/test_circle_invite_teacher.py:95]: I think it is better to say "not on"

File: open_schools_platform/organization_management/circles/tests/views/test_circle_invite_teacher.py
Comment (Ivan228777) [inline open_schools_platform/organization_management/circles/tests/views/test_circle_invite_teacher.py:107]: I think it is better to assert it with status code, that is raised by exceptions returned in our views.

File: open_schools_platform/organization_management/circles/tests/views/test_circle_invite_teacher.py
Comment (Ivan228777) [inline open_schools_platform/organization_management/circles/tests/views/test_circle_invite_teacher.py:121]: ^^

File: open_schools_platform/organization_management/circles/views.py
Comment (Ivan228777) [inline open_schools_platform/organization_management/circles/views.py:177]: no empty_exception=True?

File: open_schools_platform/organization_management/circles/views.py
Comment (Ivan228777) [inline open_schools_platform/organization_management/circles/views.py:190]: better to move status=201 to 189 line

File: open_schools_platform/organization_management/teachers/migrations/0001_initial.py
Comment (Ivan228777) [inline open_schools_platform/organization_management/teachers/migrations/0001_initial.py:84]: Are HistoricalTeacherProfile and HistoricalTeacher created automatically? Is it normal that they were created? I am just curious because I am a bit out of context)

**Review by Ivan228777 — CHANGES_REQUESTED** [review]

File: open_schools_platform/organization_management/teachers/migrations/0001_initial.py
Comment (Feazom) [reply open_schools_platform/organization_management/teachers/migrations/0001_initial.py:84]: Earlier it was added automatically, but now I corrected it and added it explicitly in the model

**Review by Ivan228777 — APPROVED** [review]
Nice, we can merge it into dev after @zavarin-michael approval.

**Review by zavarin-michael — CHANGES_REQUESTED** [review]
we must think about extensibility of query_handlers

File: open_schools_platform/organization_management/circles/services.py
Comment (zavarin-michael) [inline open_schools_platform/organization_management/circles/services.py:124]: it is ok but make it so if follow after access
it is need for new developers, they can forget to write if
you can come up better solution to not write ifs for each access

File: open_schools_platform/organization_management/circles/services.py
Comment (zavarin-michael) [inline open_schools_platform/organization_management/circles/services.py:108]: it is cool!

File: open_schools_platform/organization_management/circles/services.py
Comment (zavarin-michael) [inline open_schools_platform/organization_management/circles/services.py:103]: it's not a user_role, it's more of a user_type

File: open_schools_platform/organization_management/circles/services.py
Comment (zavarin-michael) [inline open_schools_platform/organization_management/circles/services.py:1]: you make a revolution!)
we should implement this logic in BaseQueryHandler, it would be greateful!
but i think we will do it in other branch

**Review by zavarin-michael — CHANGES_REQUESTED** [review]
change some details
great job!

File: open_schools_platform/organization_management/circles/services.py
Comment (zavarin-michael) [inline open_schools_platform/organization_management/circles/services.py:83]: I think we should make this dict:
Dict[Tuple[str, str], Tuple]
so we can write only one permission
interaction of several permission we will handle in query_handler function

File: open_schools_platform/organization_management/circles/services.py
Comment (zavarin-michael) [inline open_schools_platform/organization_management/circles/services.py:56]: we should locate to query_handler

File: open_schools_platform/organization_management/circles/services.py
Comment (zavarin-michael) [inline open_schools_platform/organization_management/circles/services.py:67]: ^^

File: open_schools_platform/organization_management/circles/services.py
Comment (zavarin-michael) [inline open_schools_platform/organization_management/circles/services.py:106]: have to be changed if we use only one permission

**Review by zavarin-michael — CHANGES_REQUESTED** [review]

File: open_schools_platform/parent_management/families/services.py
Comment (zavarin-michael) [inline open_schools_platform/parent_management/families/services.py:54]: explain how it works, please
we add attribute with FamilyQueryHandler(), why old calls of query_handler are valid
should we call it like family.query_handler.query_handler now?

File: open_schools_platform/common/services.py
Comment (zavarin-michael) [inline open_schools_platform/common/services.py:59]: please, extend docstring)
don't forget about rules, this class works only if rules has predicate_input_type_check

**Review by zavarin-michael — CHANGES_REQUESTED** [review]
cool!
finishing touch

File: open_schools_platform/parent_management/families/services.py
Comment (Feazom) [reply open_schools_platform/parent_management/families/services.py:54]: BaseQueryHandler now have __call__ function which is call query_handler

**Review by zavarin-michael — APPROVED** [review]

**Pull request merged**
